File: plot/plotgraph.cpp
//
// Plot a graph to a PNG file.
//
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <limits>
#include <new>
#include <vector>

#include <pngwriter.h>

#include "edge.hpp"
#include "point.hpp"

using namespace std;

static void get_dimensions( const vector<edge> &, point&, point& );
static string get_savename( const string& filename );
static vector<edge>& load( const string& filename );
static void plot( const vector<edge>&, const string& savefile );
static void print( vector<edge>& );

static bool debug_print = false;

int main( int argc, char **argv )
{
	int status = 0;

	cout << "Plot Graph" << endl;

	for( int i = 1; i < argc; i++ ) {
		string filename( argv[i] );
		if ( filename == "--verbose" ) {
			debug_print = true;
			continue;
		}
		if ( filename == "--quiet" ) {
			debug_print = false;
			continue;
		}

		try {
			// Load the list of edges from a file
			vector<edge>& edges = load( filename );

			if ( debug_print )
				print( edges );

			// Save the PNG image to a file
			string savename = get_savename( filename );
			if ( debug_print )
				cout << "Output file: " << savename << endl;

			// Plot the edges to a PNG image
			plot( edges, savename );

			// Clean up
			// TODO: use unique_ptr to manage memory
			delete &edges;
		}
		catch (const bad_alloc& e) {
			cout << "File         : " << filename << endl;
			cout << "Out of memory: " << e.what() << endl;
			status++;
		}
		catch (const exception& e) {
			cout << "File     : " << filename << endl;
			cout << "Exception: " << e.what() << endl;
			status++;
		}
	}

	return status;
}

static void print( vector<edge>& );
//
// Load a set of edges from a file.
//
static vector<edge>& load( const string& filename )
{
	if ( debug_print )
		cout << __FUNCTION__ << "( filename = " << filename << " )" << endl;

	ifstream in( filename );
	vector<edge> *edgelist = new vector<edge>;

	while ( true ) {
		int x1 = 0, y1 = 0, x2 = 0, y2 = 0;
		float wt;
		in >> x1 >> y1 >> x2 >> y2 >> wt;
		if ( ! in.good() )
			break;
		edgelist->emplace_back( edge( point( x1, y1 ), point( x2, y2 ) ) );
	}

	in.close();

	return *edgelist;
}

//
// Calculate the translation offset from graph-relative coordinates to
// image-relative coordinates.
//
// Move the entire image into Quadrant I
//
static point calculate_translation_offset( const point& ul, const point& lr )
{
	return point( abs( ul.get_x() ), abs( lr.get_y() ) );
}

//
// Get the upper-left and lower-right coordinate extremes from the vertices
// in a set of edges.
//
static void get_dimensions( const vector<edge> &edges, point& ul, point& lr )
{
	ul.set_x( numeric_limits<int>::max() );
	ul.set_y( numeric_limits<int>::max() );
	lr.set_x( numeric_limits<int>::min() );
	lr.set_y( numeric_limits<int>::min() );

	for ( auto& e : edges ) {
		// Find upper-left corner
		if ( e.get_a().get_x() < ul.get_x() )
			ul.set_x( e.get_a().get_x() );
		if ( e.get_a().get_y() < ul.get_y() )
			ul.set_y( e.get_a().get_y() );
		if ( e.get_b().get_x() < ul.get_x() )
			ul.set_x( e.get_b().get_x() );
		if ( e.get_b().get_y() < ul.get_y() )
			ul.set_y( e.get_b().get_y() );
		// Find lower-right corner
		if ( e.get_a().get_x() > lr.get_x() )
			lr.set_x( e.get_a().get_x() );
		if ( e.get_a().get_y() > lr.get_y() )
			lr.set_y( e.get_a().get_y() );
		if ( e.get_b().get_x() > lr.get_x() )
			lr.set_x( e.get_b().get_x() );
		if ( e.get_b().get_y() > lr.get_y() )
			lr.set_y( e.get_b().get_y() );
	}
}

//
// Translate from graph-relative coordinates to image-relative coordinates.
//
static point translate( const point& offset, const point& a )
{
	return point( a.get_x() + offset.get_x(), a.get_y() + offset.get_y() );
}

//
// Generate a PNG image plot for a set of edges and write it to the
// specified file.
//
static void plot( const vector<edge>& edges, const string& savefile )
{
	point upper_left;
	point lower_right;
	get_dimensions( edges, upper_left, lower_right );

	int width = lower_right.get_x() - upper_left.get_x();
	int height = lower_right.get_y() - upper_left.get_y();

	point offset = calculate_translation_offset( upper_left, lower_right );

	pngwriter image( width, height, 65535, savefile.c_str() );

	// Draw axes
	image.line( offset.get_x(), height, offset.get_x(), 1, 0.5, 0.5, 0.5 );
	image.line( 0, offset.get_y(), width, offset.get_y(), 0.5, 0.5, 0.5 );

	// Plot the edges
	for ( auto& e : edges ) {
		point a = translate( offset, e.get_a() );
		point b = translate( offset, e.get_b() );
		image.line( a.get_x(), a.get_y(), b.get_x(), b.get_y(), 0.0, 0.5, 0.5 );
		image.cross( a.get_x(), a.get_y(), 4, 4, 0, 0, 0 );
		image.cross( b.get_x(), b.get_y(), 4, 4, 0, 0, 0 );
	}

	// Close the image to write it to a file.
	image.close();
}

//
// Print a vector of edges
//
static void print( vector<edge>& egs )
{
	cout << "number of edges = " << egs.size() << endl;
	for ( auto& e : egs ) {
		cout << '(' << e.get_a().get_x() << ',' << e.get_a().get_y()
			<< ") -> (" << e.get_b().get_x() << ',' << e.get_b().get_y()
			<< ") wt = " << e.get_length() << endl;
	}
	cout << endl;
}

//
// Replace the data filename extension with ".png"
//
static string get_savename( const string& filename )
{
	string savename( filename );
	size_t last_dot = savename.find_last_of( '.' );
	savename.replace( last_dot, string::npos, ".png" ); 
	return savename;
}

File: plot-java/mst/PlotGraph.java
/**
 * Plot a graph to a PNG file.
 *
 * The graph is a list of edges.  An edge is a pair of vertices.
 */

package mst;

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.Iterator;
import java.util.StringTokenizer;
import java.util.Vector;

import javax.imageio.ImageIO;


public class PlotGraph
{
    /**
     * Generate a plot of each graph data file in the argument list.
     *
     * The special arguments <code>--verbose</code> and <code>--quiet</code>
     * enable or disable debugging messages.
     *
     * @param args a list of command line arguments
     *
     * The system return value is the number of files that were not
     * successfully processed.
     */
     public static void main(String[] args) {
         int status = 0;

         System.out.println("Plot Graph");

         for (int i = 0; i < args.length; i++) {
             if (args[i].equals("--verbose")) {
                 debugPrint = true;
                 continue;
             }
             if (args[i].equals("--quiet")) {
                 debugPrint = false;
                 continue;
             }

             String filename = new String(args[i]);
             try {
                 // Load the list of edges from a file
                 Vector<Edge> edges = load(filename);

                 if (debugPrint)
                     print(edges);

                 // Save the PNG image to a file
                 String savename = getSavename(filename);
                 debug("Output file: " + savename);

                 // Plot the edges to a PNG image
                 plot(edges, savename);
                 debug("\n");
             }
             catch (OutOfMemoryError e) {
                 System.err.println("File         : " + filename);
                 System.err.println("Out of memory: " + e);
                 status++;
             }
             catch (Exception e) {
                 System.err.println("File     : " + filename);
                 System.err.println("Exception: " + e);
                 status++;
             }
         }

         System.exit(status);
     }

    private static boolean debugPrint = false;

    /**
     * Print debugging messages when debugPrint is true.
     * @param s the message
     */
    private static void debug(String s) {
        if (debugPrint)
            System.out.println(s);
    }

    /**
     * Load a set of edges from a file.
     * @param filename the name of the data file
     * @return A Vector of Edges
     */
    private static Vector<Edge> load(String filename) {
        debug("load ( filename = " + filename + " )");

        Vector<Edge> edgeList = new Vector<Edge>();
        try {
            BufferedReader in = new BufferedReader(new FileReader(filename));
            String line = in.readLine();
            while (line != null && line.length() > 8) {
                StringTokenizer st = new StringTokenizer(line, "\t");
                int x1 = Integer.parseInt(st.nextToken());
                int y1 = Integer.parseInt(st.nextToken());
                int x2 = Integer.parseInt(st.nextToken());
                int y2 = Integer.parseInt(st.nextToken());
                @SuppressWarnings("unused")
                float wt = Float.parseFloat(st.nextToken());
                edgeList.add(new Edge(new Point(x1, y1), new Point(x2, y2)));
                line = in.readLine();
            }
            in.close();
        } catch (FileNotFoundException e) {
            System.out.println("FileNotFoundException: " + e);
            e.printStackTrace();
        } catch (IOException e) {
            System.out.println("IOException: " + e);
            e.printStackTrace();
        }

        return edgeList;
    }

    /**
     * Calculate the translation offset from graph-relative coordinates to
     * image-relative coordinates.
     *
     * Move the entire image into Quadrant I
     *
     * @param ul a Point in the upper left corner
     * @param lr a Point in the lower right corner
     * @param pad padding to make the image larger than the extreme points
     */
    private static Point calculateTranslationOffset(Point ul, Point lr, int pad) {
        return new Point(Math.abs(ul.getX()) + pad, Math.abs(lr.getY()) + pad);
    }

    /**
     * Get the upper-left and lower-right coordinate extremes from the vertices
     * in a set of edges.
     * @param edges a Vector of Edges
     * @param ul a ref to a Point wherein the upper left coordinate will be
     *         returned
     * @param lr a ref to a Point wherein the lower right coordinate will be
     *         returned
     */
    private static void getDimensions(Vector<Edge> edges, Point ul, Point lr) {
        ul.setX(Integer.MAX_VALUE);
        ul.setY(Integer.MAX_VALUE);
        lr.setX(Integer.MIN_VALUE);
        lr.setY(Integer.MIN_VALUE);

        Iterator<Edge> it = edges.iterator();
        while (it.hasNext()) {
            Edge e = it.next();
            // Find upper-left corner
            if ( e.getA().getX() < ul.getX() )
                ul.setX( e.getA().getX() );
            if ( e.getA().getY() < ul.getY() )
                ul.setY( e.getA().getY() );
            if ( e.getB().getX() < ul.getX() )
                ul.setX( e.getB().getX() );
            if ( e.getB().getY() < ul.getY() )
                ul.setY( e.getB().getY() );
            // Find lower-right corner
            if ( e.getA().getX() > lr.getX() )
                lr.setX( e.getA().getX() );
            if ( e.getA().getY() > lr.getY() )
                lr.setY( e.getA().getY() );
            if ( e.getB().getX() > lr.getX() )
                lr.setX( e.getB().getX() );
            if ( e.getB().getY() > lr.getY() )
                lr.setY( e.getB().getY() );
        }
    }

    /**
     * Generate a PNG image plot for a set of edges and write it to the
     * specified file.
     * @param edges A Vector of Edges
     * @param savename the name of a file to save the graph to
     */
    private static void plot(Vector<Edge> edges, String savename) {
        final int PAD = 10;
        Point upperLeft = new Point();
        Point lowerRight = new Point();

        getDimensions(edges, upperLeft, lowerRight);
        debug("upper-left : " + upperLeft.getX() + ", " + upperLeft.getY());
        debug("lower-right: " + lowerRight.getX() + ", " + lowerRight.getY());

        int width = roundUp(lowerRight.getX() - upperLeft.getX(), PAD);
        int height = roundUp(lowerRight.getY() - upperLeft.getY(), PAD);

        Point offset = calculateTranslationOffset(upperLeft, lowerRight, PAD);
        debug("offset : " + offset.getX() + ", " + offset.getY());

        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g2d = image.createGraphics();
        g2d.setBackground(Color.WHITE);
        g2d.clearRect(0, 0, width, height);

        // Draw axes
        g2d.setPaint(Color.GRAY);
        g2d.drawLine(offset.getX(), 0, offset.getX(), height);
        g2d.drawLine(0, offset.getY(), width, offset.getY());

        // Plot the edges
        Iterator<Edge> it = edges.iterator();
        while (it.hasNext()) {
            Edge e = it.next();
            Point a = translate(offset, e.getA());
            Point b = translate(offset, e.getB());

            // draw the edge
            g2d.setPaint(Color.CYAN);
            g2d.drawLine(a.getX(), a.getY(), b.getX(), b.getY());

            // draw a cross at each end point
            g2d.setPaint(Color.BLACK);
            g2d.drawLine(a.getX(), a.getY() - 2, a.getX(), a.getY() + 2);
            g2d.drawLine(a.getX() - 2, a.getY(), a.getX() + 2, a.getY());
            g2d.drawLine(b.getX(), b.getY() - 2, b.getX(), b.getY() + 2);
            g2d.drawLine(b.getX() - 2, b.getY(), b.getX() + 2, b.getY());
        }

        // Write the image to a file.
        try {
            File savefile = new File(savename);
            ImageIO.write(image, "png", savefile);
        } catch (IOException e) {
            System.out.println("IOException: " + e);
            e.printStackTrace();
        }
    }

    /**
     * Print a vector of edges (if verbose output is enabled).
     * @param edges a Vector of Edge
     */
    private static void print(Vector<Edge> edges)
    {
        if (!debugPrint)
            return;
        System.out.println("number of edges = " + edges.size());
        Iterator<Edge> it = edges.iterator();
        while (it.hasNext()) {
            Edge e = it.next();
            System.out.println( "(" +  e.getA().getX() + "," + e.getA().getY()
                + ") -> (" + e.getB().getX() + "," + e.getB().getY()
                + ") wt = " + e.getLength());
        }
        System.out.println("");
    }

    /**
     * Round the image dimension (width or height) up to the next multiple of m.
     * @param x initial dimension
     * @param m the number boundary
     * @return the adjusted dimension
     */
    private static int roundUp(int x, int m) {
        int q = (x + 2 * m) / m;
        return (q + 1) * m;
    }

    /**
     * Replace the data filename extension of ".csv" with ".png"
     * @param filename the name of the input file
     * @return the name of the output file
     */
    private static String getSavename(String filename) {
        if (!filename.endsWith(".csv"))
            throw new IllegalArgumentException("filename must end in '.csv'");
        String savename = filename.replace(".csv", ".png");
        return savename;
    }

    /**
     * Translate from graph-relative coordinates to image-relative coordinates.
     * @param offset the translation offset
     * @param a a Point to translate
     * @return the resulting Point
     */
    private static Point translate(Point offset, Point a) {
        return new Point(a.getX() + offset.getX(), a.getY() + offset.getY());
    }
}

